// @snyk/composer-lockfile-parser/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var _ = require("lodash");
var path = require("path");
var errors_1 = require("./errors");
var file_parser_1 = require("./parsers/file-parser");
var composer_parser_1 = require("./parsers/composer-parser");
function buildDepTree(lockFileContent, manifestFileContent, defaultProjectName, systemVersions, includeDev) {
    if (includeDev === void 0) { includeDev = false; }
    var lockFileJson = file_parser_1.FileParser.parseLockFile(lockFileContent);
    var manifestJson = file_parser_1.FileParser.parseManifestFile(manifestFileContent);
    if (!lockFileJson.packages) {
        throw new errors_1.InvalidUserInputError('Invalid lock file. Must contain `packages` property');
    }
    var name = manifestJson.name || defaultProjectName;
    var version = composer_parser_1.ComposerParser.getVersion(manifestJson) || '0.0.0';
    var dependencies = composer_parser_1.ComposerParser.buildDependencies(manifestJson, lockFileJson, manifestJson, systemVersions, includeDev);
    var hasDevDependencies = !_.isEmpty(manifestJson['require-dev']);
    return {
        name: name,
        version: version,
        dependencies: dependencies,
        hasDevDependencies: hasDevDependencies,
        packageFormatVersion: 'composer:0.0.1',
    };
}
exports.buildDepTree = buildDepTree;
function buildDepTreeFromFiles(basePath, lockFileName, systemVersions, includeDev) {
    if (includeDev === void 0) { includeDev = false; }
    if (!basePath) {
        throw new errors_1.InvalidUserInputError('Missing `basePath` parameter for buildDepTreeFromFiles()');
    }
    if (!lockFileName) {
        throw new errors_1.InvalidUserInputError('Missing `lockfile` parameter for buildDepTreeFromFiles()');
    }
    if (!systemVersions) {
        throw new errors_1.InvalidUserInputError('Missing `systemVersions` parameter for buildDepTreeFromFiles()');
    }
    var lockFilePath = path.resolve(basePath, lockFileName);
    var manifestFilePath = path.resolve(basePath, path.dirname(lockFilePath), 'composer.json');
    if (!fs.existsSync(lockFilePath)) {
        throw new errors_1.InvalidUserInputError("Lockfile not found at location: " + lockFilePath);
    }
    if (!fs.existsSync(manifestFilePath)) {
        throw new errors_1.InvalidUserInputError("Target file composer.json not found at location: " + manifestFilePath);
    }
    var lockFileContent = fs.readFileSync(lockFilePath, 'utf-8');
    var manifestFileContent = fs.readFileSync(manifestFilePath, 'utf-8');
    var defaultProjectName = getDefaultProjectName(basePath, lockFileName);
    return buildDepTree(lockFileContent, manifestFileContent, defaultProjectName, systemVersions, includeDev);
}
exports.buildDepTreeFromFiles = buildDepTreeFromFiles;
function getDefaultProjectName(basePath, lockFileName) {
    return path.dirname(path.resolve(path.join(basePath, lockFileName))).split(path.sep).pop();
}
