// snyk/dist/lib/feature-flags.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const request = require("./request");
const snyk = require("."); // TODO(kyegupov): fix import
const config = require("./config");
const common_1 = require("./snyk-test/common");
function isFeatureFlagSupportedForOrg(featureFlag, org) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield request({
            method: 'GET',
            headers: {
                Authorization: `token ${snyk.api}`,
            },
            qs: common_1.assembleQueryString({ org }),
            url: `${config.API}/cli-config/feature-flags/${featureFlag}`,
            gzip: true,
            json: true,
        });
        return response.body;
    });
}
exports.isFeatureFlagSupportedForOrg = isFeatureFlagSupportedForOrg;
